### ISSUE TITLE
Remove `.data`

### ISSUE THREAD

**Issue by albanD** [pytorch/pytorch#30987]
Even though it is not documented, many users still use it. And it leads to many bugs in user code.
So we should remove it completely to prevent this.
The expected steps are:
- [ ] Add a new api to make a shallow copy that does not share version? Or a cat + unbind function that does not share version counter if that is enough.
- [ ] Remove the use of .data in all our internal code
- [ ] Add warning to every call to `.data`
- [ ] Actually remove `.data` ?

**Comment by vadimkantorov** [comment]
I currently have to use `.data` to circumvent version tracking for implementing invertible inplace ops (InplaceBatchNorm + InplaceLeakyRelu). I wish there was a way around that.

**Comment by albanD** [comment]
That would be the first point. If you just want to get a copy that does not share version counter. But that new api will only do this. And won't have all the other side-effects of `.data`. That should fit your use case right?

**Comment by vadimkantorov** [comment]
Yes I think so! :)

**Comment by ZisIsNotZis** [comment]
I'm curious is `.data` "redundant" or "internal". When building custom logger, optimizer and etc, I often use `a.data = b.data` where `b` was a clone of `a`. Literally, this should mean "attach the data of `b` back to `a` without copying", but I'm not sure did pytorch actually do this.

One good thing to have `.data` is that you can do these kind of attaching/detaching conveniently (while you cannot attach an `np.ndarray` to another one without actually replacing the array object). Another good thing is a short-hand notation for `torch.no_grad`

**Comment by ezyang** [comment]
cc @colesbury

**Comment by albanD** [comment]
@ZisIsNotZis it is "historical". `.data` was the Tensor wrapped by a Variable. But since Variable disappeared. It does not make sense anymore. It allows you to do too many things and in general shoot you in the foot.
For your use case, you should use `a.set_(b)` which is a specialized function to set a Tensor in another inplace. Or just `a = b` if you don't need to change `a` inplace.

**Comment by henrique** [comment]
Hey @albanD 
How does it affect the p.grad.data and p.data used by the optimisers?
e.g. https://github.com/pytorch/pytorch/blob/master/torch/optim/adam.py#L64
Just in case I'm writing my own optim ;) Cheers

**Comment by albanD** [comment]
Hi,

For optimizers, the simplest solution is to remove all the `.data` and add the `@torch.no_grad()` decorator to the `step()` function.

**Comment by gchanan** [comment]
You should document every thing `.data` currently allows you to do, to make sure we have a preferred workaround.  For example, one thing I don't see listed here is changing the dtype of a tensor in place.

**Comment by albanD** [comment]
Currently `.data` have few use. The main ones are below:
- `y = x.data` is often used to get a new Tensor that does not share the history of `x` but looks at the same memory. This should be replaced with `y = x.detach()`
- `x.data.copy_(foo)` is used to change the content of `x` without the autograd knowing about it. this should be preferably replaced by 
```
with torch.no_grad():
  x.copy_(foo)
```
- `y = x.data` is sometimes used (in `torch/cuda/` for example) to get a new Tensor that share the same data as `x` but does not share the version counter (and so has a different behavior wrt inplace operations). This is the use for which we want to add a new api.
- `x.data = y` is sometimes used (in nn.Module `_apply` function) to mutate the Tensor `x` inplace. This changes the properties of the Tensor x (type, storage, layout etc) inplace. We should disallow such modification altogether.

**Comment by vadimkantorov** [comment]
about last point: sometimes reinterpret is useful and can be made explicit: https://github.com/pytorch/pytorch/issues/29013

**Comment by albanD** [comment]
But that last point would not allow you to do this as it only populates with all the properties (including storage) of the other Tensor. So it makes a "copy" of the metadata `y` into `x`.
But I agree that a reinterpret would be useful feature !

**Comment by lanpa** [comment]
@albanD I just checked that the only `.data` used in tensorboard is the following, which is guarded by a type check.


```python
def _prepare_pytorch(x):
    if isinstance(x, torch.autograd.Variable):
        x = x.data
    x = x.cpu().numpy()
    return x
```

**Comment by albanD** [comment]
@lanpa thanks for looking into this !
Given that `torch.autograd.Variable` do not exist anymore (it's litteratlly an empty class [here](https://github.com/pytorch/pytorch/blob/14593f077f9cb248cca85fd18b598d14c47d5d4e/torch/autograd/variable.py#L10)) this check is a bit out-dated.
But if you send a PR (add me as a reviewer) to replace it with `x = x.detach().cpu().numpy()`, that will work.

**Comment by ezyang** [comment]
Is it really necessary to detach in the code above?

**Comment by albanD** [comment]
If a Tensor that requires gradient is passed to this function, the call to `.numpy()` will fail.
Given the original code, I guessed a Tensor that requires gradient can be passed. If it is never the case, then you can ignore the detach() and simply remove the `if` statement indeed.

**Comment by Enealor** [comment]
I noticed that there is renewed interest in removing `.data`. If work hasn't already begun on this, I plan on working on removing its usage in the optimizers.

**Comment by Enealor** [comment]
I started removing `.data` from the optimizers (see #33640). As a quick note, using the `@torch.no_grad` decorator for the `step` method can fail as the `closure` function may call `backward`. If the decorator is used, then `closure` will error, as I found when running the tests. This can be worked around by the end-user with `torch.enable_grad()`, so it's only an issue for the optimizers included with PyTorch.

I am also running into a further problem with my current strategy for removing `.data` from the optimizers. It seems to fail in the distributed case specifically, and I could use some advice on how to fix it or further test it. The shallow copy is complaining of in-place modifications, and I think it's related to `torch.no_grad()` being thread-local.

**Comment by albanD** [comment]
Hi,

Yes, given all the confusion it creates for new users and the danger it is, we're trying to remove it !
I'll answer on the PR.

But one of the dangerous side-effect of using `.data` is that it does not share the version counter of the original Tensor. And so the inplace verification we have in the autograd engine do not work.
Removing the `.data` adds these check back (which is good), but in some cases they are too strict. See the first point in the PR about adding a new function to do a shallow copy without sharing the version counter for advanced user.

**Comment by rodrigoberriel** [comment]
FYI, there are two mentions in the docs that remain there:
- [`onnx.rst`](https://pytorch.org/docs/master/onnx.html#avoid-tensor-data):

    ```text
    ./source/onnx.rst:Avoid Tensor.data
    ./source/onnx.rst:Using the Tensor.data field can produce an incorrect trace and therefore an incorrect ONNX graph.
    ./source/onnx.rst:`remove Tensor.data entirely <https://github.com/pytorch/pytorch/issues/30987>`_).
    ```

- [`autograd.rst`](https://pytorch.org/docs/master/autograd.html#variable-deprecated):

    ```
    ./source/autograd.rst:    - ``var.data`` is the same thing as ``tensor.data``.
    ```

I guess they'll be removed only after `.data` is finally removed or, at least, after a warning is added.

**Comment by Darkev77** [comment]
How shall one "manually" update the parameters (via SGD for example) without using `.data`? An example scenario:
`        grad = torch.autograd.grad(loss, m.parameters())
        for p, g in zip(m.parameters(), grad):
            p.data -= LR * g
`
Without `p.data` you won't be able to access the parameter's weights directly and update them since they're considered as leaf Variables, and hence can't apply in-place operations. What's the work around for this situation?

**Comment by Darkev77** [comment]
How shall one "manually" update the parameters (via SGD for example) without using `.data`? An example scenario:
> 
        grad = torch.autograd.grad(loss, m.parameters())
        for p, g in zip(m.parameters(), grad):
            p.data -= LR * g
Without `p.data` you won't be able to access the parameter's weights directly and update them since they're considered as leaf Variables, and hence can't apply in-place operations. What's the work around for this situation?
